### ISSUE TITLE
swin-L模型训练时loss为nan

### ISSUE THREAD

**Issue by CloudWarGit** [Oneflow-Inc/vision#237]
您好，在用swin_large_patch4_window7_224_in22k_to_1k模型训练的时候，loss函数输出为nan

**Comment by s14n** [comment]
你好，您是使用swin-large进行finetune吗？您是使用imagenet还是自有的数据？您是参考microsoft/Swin-Transformer设置的配置还是自己设计的参数？是否方便提供更多的信息，我们来尝试复现一下您提到的问题，谢谢。

**Comment by Ldpe2G** [comment]
@CloudWarGit 你可以用 https://github.com/Oneflow-Inc/vision/blob/main/projects/classification/ddp_training.sh#L6
 这个Project 来测试，我这边试了一下 swin_large_patch4_window7_224_in22k_to_1k 没出现 loss nan 的问题

**Comment by yuanms2** [comment]
> @CloudWarGit 你可以用 https://github.com/Oneflow-Inc/vision/blob/main/projects/classification/ddp_training.sh#L6 这个Project 来测试，我这边试了一下 swin_large_patch4_window7_224_in22k_to_1k 训 imagenet 没出现 loss nan 的问题

好的
尽管这个配置没有问题，最好把CloudWarGit那边为什么会有问题弄清楚

**Comment by CloudWarGit** [comment]
> @CloudWarGit 你可以用 https://github.com/Oneflow-Inc/vision/blob/main/projects/classification/ddp_training.sh#L6
> 这个Project 来测试，我这边试了一下 swin_large_patch4_window7_224_in22k_to_1k 训 imagenet 没出现 loss nan 的问题

是用的这个，可以提供一个完整的配置么？我的配置如下：

`DATA:
  BATCH_SIZE: 128
  DATASET: imagenet
  DATA_PATH: /imagenet/imagenet/
  IMG_SIZE: 224
  INTERPOLATION: bicubic
  ZIP_MODE: False
  CACHE_MODE: "part"
  PIN_MEMORY: True
  NUM_WORKERS: 8

MODEL:
  DROP_PATH_RATE: 0.2

TRAIN:
  EPOCHS: 30
  WARMUP_EPOCHS: 5
  WEIGHT_DECAY: 1e-8
  BASE_LR: 2e-05
  WARMUP_LR: 2e-08
  MIN_LR: 2e-07

  LR_SCHEDULER:
    NAME: cosine
    MILESTONES: None
`

另外请问下这个模型可以用静态图Graph来跑么？

**Comment by Ldpe2G** [comment]
我没改配置，就改了 imagenet 数据路径和 bash 脚本里 模型的名字

**Comment by Ldpe2G** [comment]
> > @CloudWarGit 你可以用 https://github.com/Oneflow-Inc/vision/blob/main/projects/classification/ddp_training.sh#L6
> > 这个Project 来测试，我这边试了一下 swin_large_patch4_window7_224_in22k_to_1k 训 imagenet 没出现 loss nan 的问题
> 
> 是用的这个，可以提供一个完整的配置么？我的配置如下：
> 
> `DATA: BATCH_SIZE: 128 DATASET: imagenet DATA_PATH: /imagenet/imagenet/ IMG_SIZE: 224 INTERPOLATION: bicubic ZIP_MODE: False CACHE_MODE: "part" PIN_MEMORY: True NUM_WORKERS: 8
> 
> MODEL: DROP_PATH_RATE: 0.2
> 
> TRAIN: EPOCHS: 30 WARMUP_EPOCHS: 5 WEIGHT_DECAY: 1e-8 BASE_LR: 2e-05 WARMUP_LR: 2e-08 MIN_LR: 2e-07
> 
> LR_SCHEDULER: NAME: cosine MILESTONES: None `
> 
> 另外请问下这个模型可以用静态图Graph来跑么？

flowvision 的模型没考虑 多卡的graph，如果要跑只能跑单卡 @CloudWarGit

**Comment by s14n** [comment]
你好，您是使用的预训练模型（pretrained=True）进行finetune的吗？ @CloudWarGit

**Comment by CloudWarGit** [comment]
> > > @CloudWarGit 你可以用 https://github.com/Oneflow-Inc/vision/blob/main/projects/classification/ddp_training.sh#L6
> > > 这个Project 来测试，我这边试了一下 swin_large_patch4_window7_224_in22k_to_1k 训 imagenet 没出现 loss nan 的问题
> > 
> > 
> > 是用的这个，可以提供一个完整的配置么？我的配置如下：
> > `DATA: BATCH_SIZE: 128 DATASET: imagenet DATA_PATH: /imagenet/imagenet/ IMG_SIZE: 224 INTERPOLATION: bicubic ZIP_MODE: False CACHE_MODE: "part" PIN_MEMORY: True NUM_WORKERS: 8
> > MODEL: DROP_PATH_RATE: 0.2
> > TRAIN: EPOCHS: 30 WARMUP_EPOCHS: 5 WEIGHT_DECAY: 1e-8 BASE_LR: 2e-05 WARMUP_LR: 2e-08 MIN_LR: 2e-07
> > LR_SCHEDULER: NAME: cosine MILESTONES: None `
> > 另外请问下这个模型可以用静态图Graph来跑么？
> 
> flowvision 的模型没考虑 多卡的graph，如果要跑只能跑单卡 @CloudWarGit

哦好的，那如果想改成多卡的graph，需要做哪些改造呢？我想用ddp来跑多卡的graph

**Comment by CloudWarGit** [comment]
> 你好，您是使用的预训练模型（pretrained=True）进行finetune的吗？ @CloudWarGit

没有，直接跑的

**Comment by CloudWarGit** [comment]
> > > @CloudWarGit 你可以用 https://github.com/Oneflow-Inc/vision/blob/main/projects/classification/ddp_training.sh#L6
> > > 这个Project 来测试，我这边试了一下 swin_large_patch4_window7_224_in22k_to_1k 训 imagenet 没出现 loss nan 的问题
> > 
> > 
> > 是用的这个，可以提供一个完整的配置么？我的配置如下：
> > `DATA: BATCH_SIZE: 128 DATASET: imagenet DATA_PATH: /imagenet/imagenet/ IMG_SIZE: 224 INTERPOLATION: bicubic ZIP_MODE: False CACHE_MODE: "part" PIN_MEMORY: True NUM_WORKERS: 8
> > MODEL: DROP_PATH_RATE: 0.2
> > TRAIN: EPOCHS: 30 WARMUP_EPOCHS: 5 WEIGHT_DECAY: 1e-8 BASE_LR: 2e-05 WARMUP_LR: 2e-08 MIN_LR: 2e-07
> > LR_SCHEDULER: NAME: cosine MILESTONES: None `
> > 另外请问下这个模型可以用静态图Graph来跑么？
> 
> flowvision 的模型没考虑 多卡的graph，如果要跑只能跑单卡 @CloudWarGit

您好，我再确认下是graph不支持多卡，还是这个模型不支持多卡的graph？

**Comment by Ldpe2G** [comment]
@CloudWarGit  graph 支持多卡，如果要训 swin graph 多卡，可以用我们的 libai 库
https://github.com/Oneflow-Inc/libai/blob/main/configs/swin_imagenet.py

**Comment by CloudWarGit** [comment]
> @CloudWarGit graph 支持多卡，如果要训 swin graph 多卡，可以用我们的 libai 库
> https://github.com/Oneflow-Inc/libai/blob/main/configs/swin_imagenet.py

感谢，用libai跑起来了，但是有个问题，用libai的默认配置跑，单机单卡跑的时候这个模型吞吐可以到600+，但是单机8卡只能跑到2400+多，线性加速比有点低。默认参数是不哪里可以优化下？

**Comment by CloudWarGit** [comment]
> > > @CloudWarGit graph 支持多卡，如果要训 swin graph 多卡，可以用我们的 libai 库
> > > https://github.com/Oneflow-Inc/libai/blob/main/configs/swin_imagenet.py
> > 
> > 
> > 感谢，用libai跑起来了，但是有个问题，用libai的默认配置跑，单机单卡跑的时候这个模型吞吐可以到600+，但是单机8卡只能跑到2400+多，线性加速比有点低。默认参数是不哪里可以优化下？另外可以给一个finetuning的模型么？
> 
> 请教一下更多信息： 你的显卡型号是什么？ 显卡之间是PCIe，还是NVLink?
> 
> 您是运行的哪个脚本？ 使用的8卡的config文件是什么，并行的设置是什么，是数据并行，还是模型并行，还是带有流水并行？
> 
> 对于有的并行模式8卡加速比是4可能已经比较高了，需要看到更多的配置信息才能诊断这个结果是否合理
> 
> 您提供更多硬件配置信息之后，oneflow 同事也可以和自己跑的结果做一下对比

A100，NVLink，运行的是tools/train.sh，配置文件https://github.com/Oneflow-Inc/libai/blob/main/configs/swin_imagenet.py。
我直接bash tools/train.sh tools/train_net.py 跑的，我理解是数据并行吧，用pytorch跑swin-transformer，线性加速比可以到0.95+

**Comment by Ldpe2G** [comment]
> 另外要跑large的话，是不只要改下swin_imagenet.py里的`from .common.models.swin.swin_large_patch4_window7_224 import model`就行了吧？

是的

**Comment by Ldpe2G** [comment]
> 我直接bash tools/train.sh tools/train_net.py 跑的

可以试试在配置里设置 graph ，可以参考: https://github.com/Oneflow-Inc/libai/blob/main/configs/swin_cifar100.py#L58

@CloudWarGit

**Comment by CloudWarGit** [comment]
> > 我直接bash tools/train.sh tools/train_net.py 跑的
> 
> 可以试试在配置里设置 `graph.enabled = True`，可以参考: https://github.com/Oneflow-Inc/libai/blob/main/configs/swin_cifar100.py#L58
> 
> @CloudWarGit

默认是开启的
<img width="682" alt="image" src="https://user-images.githubusercontent.com/10798768/176152382-d140de6d-7d66-4539-a6dc-d3efa2278af7.png">

`graph.auto_parallel.enabled` 需要改成true么？

**Comment by Ldpe2G** [comment]
试试打开amp `train.amp.enabled = True`

https://github.com/Oneflow-Inc/libai/blob/main/configs/swin_cifar100.py#L54

**Comment by CloudWarGit** [comment]
> 试试打开amp `train.amp.enabled = True` @CloudWarGit
> 
> https://github.com/Oneflow-Inc/libai/blob/main/configs/swin_cifar100.py#L54

这个也是开启的，我用的这个配置：[https://github.com/Oneflow-Inc/libai/blob/main/configs/swin_imagenet.py](https://github.com/Oneflow-Inc/libai/blob/main/configs/swin_imagenet.py%E3%80%82)

**Comment by Ldpe2G** [comment]
我这边在v100上测了下，[默认配置](https://github.com/Oneflow-Inc/libai/blob/main/configs/swin_imagenet.py)下 8卡加速比4倍左右，还测了下官方的 swin, 8卡加速比 6~7倍
